pkg/token/serviceprincipaltoken.go

```diff
@@ -43,14 +60,47 @@ func (p *servicePrincipalToken) Token() (adal.Token, error) {
 	callback := func(t adal.Token) error {
 		return nil
 	}
-	spt, err := adal.NewServicePrincipalToken(
-		p.oAuthConfig,
-		p.clientID,
-		p.clientSecret,
-		p.resourceID,
-		callback)
-	if err != nil {
-		return emptyToken, fmt.Errorf("failed to create service principal token: %s", err)
+
+	var (
+		spt       *adal.ServicePrincipalToken
+		callbacks []adal.TokenRefreshCallback
+		err       error
+	)
+
+	if p.clientSecret != "" {
+		spt, err = adal.NewServicePrincipalToken(
+			p.oAuthConfig,
+			p.clientID,
+			p.clientSecret,
+			p.resourceID,
+			callback)
+		if err != nil {
+			return emptyToken, fmt.Errorf("failed to create service principal token using secret: %s", err)
+		}
+	} else if p.clientCert != "" {
+		certData, err := ioutil.ReadFile(p.clientCert)
+		if err != nil {
+			return emptyToken, fmt.Errorf("failed to read the certificate file (%s): %v", p.clientCert, err)
+		}
+
+		// Get the certificate and private key from pfx file
+		cert, rsaPrivateKey, err := decodePkcs12(certData, "")
+		if err != nil {
+			return emptyToken, fmt.Errorf("failed to decode pkcs12 certificate while creating spt: %v", err)
+		}
+
+		callbacks = append(callbacks, callback)
+
+		spt, err = adal.NewServicePrincipalTokenFromCertificate(
+			p.oAuthConfig,
+			p.clientID,
+			cert,
+			rsaPrivateKey,
+			p.resourceID,
+			callbacks...)
```

Comment:
```suggestion
			callback)
```
would this work? like the SP using secret above?

Reply:
done